### ISSUE TITLE
Remove IOptionsFactory new constraint

### ISSUE THREAD

**Issue by NinoFloris** [aspnet/Extensions#1463]
Why was IOptionsFactory ever released with a `new()` constraint on the interface?!
It kills a lot of useful other construction patterns where you might not have an empty constructor for a good reason, (which is why you'd want a factory interface in the first place) 

And it also makes it painful to use with F# records (which don't have an empty constructor).

I'm not sure why this constraint wasn't put onto the `IServiceCollection.Configure` method type argument?

Could anybody explain?

**Comment by HaoK** [comment]
Well the original IOptions had a `new()` constraint, and the default implementation of the OptionsFactory is using new().  The factory wasn't really meant changing how TOptions instances are created (which is always done via new), but more for applying all of the configurations to the instance.

**Comment by HaoK** [comment]
Its main job is to apply these:

```
OptionsFactory(IEnumerable<IConfigureOptions<TOptions>> setups, IEnumerable<IPostConfigureOptions<TOptions>> postConfigures, IEnumerable<IValidateOptions<TOptions>> validations)
```

**Comment by NinoFloris** [comment]
Would there still be some room to re-evaluate this? 

Like either removing the constraint from the interface and moving it up to the Configure method.
Or possibly detect another, unconstrained, factory interface during options creation?

You would make a lot of people with uncommon use cases happy with such a change, and the F# community would love it too.

**Comment by NinoFloris** [comment]
@haok ping.

I'm willing to drive this break too, investigate breaking places, fix the code, write up a breaking change issue etcetera. But 3.0 is the best time to remove this constraint if possible.

**Comment by HaoK** [comment]
Feel free to give it a shot so we can see what it looks like, but no guarantees this will be something we take.

**Comment by HaoK** [comment]
Addressed by https://github.com/aspnet/Extensions/pull/2169

**Comment by HaoK** [comment]
@anurse this is considered a 5.0 change since it was merged to master right?

**Comment by anurse** [comment]
Correct. I put it in the placeholder `5.0.0-alpha1` milestone (just to track that it was in before any public preview releases), which isn't a committment to any specific release, just a placeholder until the schedule starts to take shape :).

**Comment by NinoFloris** [comment]
@anurse did you get more info on backporting to 3.0?

**Comment by anurse** [comment]
I'll run it up the flagpole but a change like this to a public API will be challenging. Do you have specific scenarios in production applications that don't work **and** for which there is no workaround? That's probably what we're going to need to make an API change (especially since I believe this is a breaking change to an interface) in 3.0.

**Comment by NinoFloris** [comment]
>Do you have specific scenarios in production applications that don't work and for which there is no workaround?
We don't, possibly people want to use types that they don't control and which don't have an empty constructor.

>(especially since I believe this is a breaking change to an interface) 
I always thought removing constraints was non breaking?

>but a change like this to a public API will be challenging
That saddens me a bit, I would have loved to have it in for 3.0. But, and I think this is a fair point of critique, you all went rather radio silent on this repo for months when 3.0 was still open for changes.

**Comment by anurse** [comment]
> I always thought removing constraints was non breaking?

Yep, I suspected that was the case too but wasn't sure. I've confirmed it is indeed non-breaking. Removing the constraint is generally fine, my concern was how it would affect implementations. It looks like it's safe, after [exploring the scenarios in SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0ATEBqAPgenwAIA5GAdyIEsA7AFxigDMBDMGNI8qugCyMg0AznSgtadIrAC2EAG4wMAWABQAAQDM1eo1bsiASQDyABzpUIwgGJs60AJ4AeACoA+VQG9VAX1WrCREYANhgCEBgwRHwskjA0GEJEAK4mRGoADFG8kTQU2gzMbDCqmmkATIFmFta2Di6uRCCGpuaWQjZgdlBOblzZsETOTbnkABQAlJ4+fioBAMIQ0iZUQTCJRgDSAHRAA)

> That saddens me a bit, I would have loved to have it in for 3.0. But, and I think this is a fair point of critique, you all went rather radio silent on this repo for months when 3.0 was still open for changes.

Yep, definitely a fair observation. We're a small team with a lot of code to manage and things get prioritized (and deprioritized) as part of that. I don't disagree that this repo has been quiet for a while though, and that's something we're looking to improve.

Regardless of how we got here, we still have to balance the risk of any change in 3.0 (there's always non-zero risk) against the customer value it delivers. A lot of that is driven by having a clear understanding of customer scenarios that are blocked by the issue and the value of taking the backport.
